### PR TITLE
Decentralized support

### PR DESCRIPTION
**Pull request by JibrilExe**
Making PPOTrainer generic, so we can choose between centralized, arm or segment lvl with a config param.

Started with adapting the obs_dict convert function, it was the perfect point that already gets the padded obs dict.
Depending on morphology, here we split the env state.

All other code that expected (num_envs, flattened_obs_shape) will be changed to be capable of using (num_envs, agents, flattened_obs_shape). In the centralized case, agents will be 1.

Actual message passer function will also be implemented here, it will come down to using a mask based on adjacency list, passing around hidden state, aggregation, repeat, return result, and with jax this should be vectorized over all agents and then over all envs..

Finally we have to decide a way to build adjacency, but this can be made as simple as we want it, for starters we can just hardcode based on morphology. We do need a way to make out if a MLP is center ring or in the arm.

### CODE REVIEW

File: configs/experiment/long_2arm.yaml
Comment (tdpeuter) [inline configs/experiment/long_2arm.yaml:1]: Sure, but I would rather override the experiment name on cli

File: scripts/hpc/train.pbs [hunk 1]
```diff
@@ -66,7 +66,7 @@ fi
 # Run training using Hydra overrides
 python scripts/train.py \
     hydra.run.dir="$SCRATCH_RUNDIR" \
-    ppo=stable \
+    ppo=smoke_test \
```
Comment (tdpeuter) [inline scripts/hpc/train.pbs]:
```suggestion
    ppo=stable \
```
I would try to keep the HPC setup for long experiments. Of course you can modify the script locally, but I wouldn't push them.

File: tests/test_adjacency.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+def test_ring():
+    adj = build_adjacency([4, 4, 4, 4, 4], MorphMode.RING)
+
+    assert adj.shape == (5, 5)
+    assert_symmetric(adj)
+
+    # each node should connect to itself + 2 neighbors
+    for i in range(5):
+        assert adj[i, i] == 1
+        assert jnp.sum(adj[i]) == 3
```
Comment (tdpeuter) [inline tests/test_adjacency.py]:
```suggestion
    # each node should connect to itself + 2 neighbors
    for node in range(5):
        assert adj[node, node] == 1
        assert jnp.sum(adj[node]) == 3
        neighbor1 = (node - 1) % 5
        neighbor2 = (node + 1) % 5
        assert adj[neighbor1, node] == 1
        assert adj[node, neighbor2] == 1  # Symmetrical
```
Explicitly testing of the actual neighbors are connected, instead of only summing. Seems to work locally.

File: src/brittle_star_project/trainers/PPOTrainer.py [hunk 3]
```diff
@@ -407,7 +497,25 @@ def __init__(
...
+
+        def apply_critic(p, x):
+            return apply_shared(self.critic, p, x)
+
+        def apply_feature(p, x):
+            return apply_shared(self.feature_extractor, p, x)
+
+        def apply_message_passer(p, x):
+            assert self.message_passer is not None
+            return jax.vmap(lambda x_in: self.message_passer.apply(p, x_in))(x)
```
Comment (tdpeuter) [inline src/brittle_star_project/trainers/PPOTrainer.py:532]:
```suggestion
        def apply_message_passer(p, x):
        	if self.message_passer is None:
        		return x
            return jax.vmap(lambda x_in: self.message_passer.apply(p, x_in))(x)
```
To allow centralized morphologies as well? This way we don't have to check at the call site. (Tested locally, now allows to train centralized models again.)

File: src/brittle_star_project/environment/obs_processing.py [hunk 4]
```diff
@@ -51,41 +101,127 @@ def _normalize_features(obs: dict) -> dict:
...
+        total = 0
+        for k, v in obs.items():
+            if hasattr(v, "shape"):
+                size = v.size
+                logger11.info(f"[RAW] {k}: shape={v.shape}, size={size}")
+                total += size
+            else:
+                logger11.info(f"[RAW] {k}: non-array")
+
+        logger11.info(f"[RAW TOTAL FEATURES]: {total}")
```
Comment (tdpeuter) [inline src/brittle_star_project/environment/obs_processing.py]:
```suggestion
        logger11.debug(f"[RAW TOTAL FEATURES]: {total}")
```
Please use debug for this and subsequent "RAW" related logs, to declutter the output.

**Review by tdpeuter — CHANGES_REQUESTED** [review]
Did not have time yet to look at everything in this PR (to help with the bugs), but here is my intermediate review. (18/24 files viewed - note to self: still have to look at environment/*, MLPs/*, PPOTrainer, ppo)

Also, what does `logger11` instead of just `logger` mean?

File: src/brittle_star_project/trainers/PPOTrainer.py (hunk 3 above)
Comment (RobinMeersman) [reply src/brittle_star_project/trainers/PPOTrainer.py:532]: Applied at call site, as mentioned after SCRUM

File: src/brittle_star_project/environment/obs_processing.py (hunk 4 above)
Comment (RobinMeersman) [reply src/brittle_star_project/environment/obs_processing.py]: Done

File: src/brittle_star_project/trainers/PPOTrainer.py [hunk 5]
```diff
@@ -423,33 +536,137 @@ def _init_random(self):
 
     def _init_agent(self):
         self.logger.info("[AGENT]: Initializing agent...")
-
+        agent_indices = [0, 1, 2, 3, 4]
+        match self.morph_mode:
+            case MorphMode.CENTRALIZED:
+                needed_copies = 1
```
Comment (tdpeuter) [inline src/brittle_star_project/trainers/PPOTrainer.py:562]:
```suggestion
        match self.morph_mode:
            case MorphMode.CENTRALIZED:
                needed_copies = 1
                agent_indices = [0]
```
Makes it more explicit in my opinion. I assume this does not break anything?

File: src/brittle_star_project/trainers/PPOTrainer.py [hunk 6]
```diff
@@ -219,6 +258,26 @@ def _dont(_):
...
+
+def apply_shared(net, params, x):
+    # x: (batch, nodes, feat)
+    # If the critic expects a single vector per environment:
+    batch_size = x.shape[0]
+    x_flattened = x.reshape(batch_size, -1)
+    return jax.vmap(lambda xi: net.apply(params, xi))(x_flattened)
+
+
+# TODO: update to work with extra dimension + message passing
```
Comment (tdpeuter) [inline src/brittle_star_project/trainers/PPOTrainer.py]:
```suggestion
```

Done?

File: src/brittle_star_project/environment/obs_processing.py [hunk 7]
```diff
@@ -51,41 +101,127 @@ def _normalize_features(obs: dict) -> dict:
...
             elif key in _SEGMENT_SCALED_KEYS:
-                padded_arr = jnp.zeros(padding_masks["target_size_1x"], dtype=arr.dtype)
-                padded[key] = padded_arr.at[padding_masks["mask_1x"]].set(arr)
+                target_size = padding_masks["target_size_1x"]
+                out = jnp.zeros((agent_count, target_size), dtype=arr.dtype)
+                out = out.at[:, : arr.shape[1]].set(arr)
+                padded[key] = out
             else:
                 padded[key] = arr
         return padded
```
Comment (tdpeuter) [inline src/brittle_star_project/environment/obs_processing.py]: This function is not getting used any more, so I guess we should remove it?

File: src/brittle_star_project/environment/obs_processing.py [hunk 8]
```diff
@@ -18,9 +24,53 @@
...
+    ordered_keys = frozenset(
+        [
+            "disk_z_tilt",
+            "joint_actuator_force",
+            "joint_position",
+            "joint_velocity",
+            "robot_direction_to_target",
+            "segment_contact",
+        ]
+    )
```
Comment (tdpeuter) [inline src/brittle_star_project/environment/obs_processing.py:70]:
```suggestion
    ordered_keys = (
        "disk_z_tilt",
        "joint_actuator_force",
        "joint_position",
        "joint_velocity",
        "robot_direction_to_target",
        "segment_contact",
    )
```

Based on what I could find online, [frozensets are not ordered](https://docs.python.org/3.15/library/stdtypes.html#frozenset).

> Since sets only define partial ordering (subset relationships), the output of the [list.sort()](https://docs.python.org/3.15/library/stdtypes.html#list.sort) method is undefined for lists of sets.

I think we should keep this as a simpler iterable, such that we can safely iterate over them.

Making this O(1) for a list of 6 keys is premature optimization anyway 👀

File: src/brittle_star_project/environment/obs_processing.py [hunk 9]
```diff
@@ -51,41 +101,127 @@ def _normalize_features(obs: dict) -> dict:
...
+            # -------- SEGMENTS --------
+            if key in _SEGMENT_SCALED_KEYS:
+                per_agent = []
+
+                for i, agent_id in enumerate(agent_indices):
+                    idx = segment_indices[i]
+                    taken = jnp.take(arr, idx, axis=0)  # (segs, ...)
+                    logger11.debug(f"WHY {taken.shape}")
+                    # pad to 4
+                    pad_len = 4 - taken.shape[0]
```
Comment (tdpeuter) [inline src/brittle_star_project/environment/obs_processing.py]: I don't like these magic values (there is an eight somewhere in here as well). Better option would be to use the computed padding masks, or make sure to derive the features from anywhere. Then, in the future, when the morphology might be changed (probably not in this project), these values don't have to be updated and hardcoded again.

**Review by tdpeuter — CHANGES_REQUESTED** [review]
Thanks for the hard work, Cedric and Robin.

I added some notes and requested changes. (I did go through everything this time.)

I am also soon pushing a commit that cleans some other things up and hopefully fixes the failing test. (We will discuss over Discord, since you are working on things as well.)

File: src/brittle_star_project/environment/obs_processing.py (hunk 8 above)
Comment (tdpeuter) [reply src/brittle_star_project/environment/obs_processing.py:70]: You can also use a copy of the list in a frozenset and a copy in a normal tuple for example, then use them in the right places.

But you discussed over Discord that we will just use `sorted()` in the for loops. That is fine 👍

**Comment by RobinMeersman** [comment]
- [x]    🔴 CRITICAL    Centralized mode never pads — _split_to_agents just reshapes, no padding    Training + Simulation
- [x]    🔴 CRITICAL    _pad_features is defined but never called from _process_single    Training + Simulation
- [x]    🟠 MODERATE    Decentralized jnp.take(arr, agent_id, axis=0) takes a single scalar, not an arm slice    Training (decentralized)
- [x]    🟠 MODERATE    ordered_keys is a frozenset — non-deterministic feature ordering between runs    Training + Simulation
- [x]    🟠 MODERATE    simulate.py missing required args (num_arms, needed_copies) — would crash    Simulation
- [x]    🟡 MINOR    build*_indices results computed but never used (dead code)    N/A
- [ ]    🟡 MINOR    _build_joint_indices doesn't advance start for excluded arms    Would be wrong if wired in

File: src/brittle_star_project/environment/obs_processing.py (hunk 9 above)
Comment (RobinMeersman) [reply src/brittle_star_project/environment/obs_processing.py]: I've put them in simple variables for now describing what they mean.

File: src/brittle_star_project/trainers/PPOTrainer.py (hunk 5 above)
Comment (RobinMeersman) [reply src/brittle_star_project/trainers/PPOTrainer.py:562]: I am ignoring this as I do not quite understand how this is handled

File: scripts/train.py [hunk 10]
```diff
@@ -41,6 +42,7 @@ def main(dict_cfg: DictConfig):
         base_dir=os.path.dirname(run_dir),
     )
     logger = get_logger()
+    logger.set_level(logging.DEBUG)
```
Comment (tdpeuter) [inline scripts/train.py]: Might want to remove this to avoid excessive logging during "final" training.

Which makes me think: is there currently no way to override the verbosity level without modifying source code? I mean, not that it is important any more in this stage of the project.

File: src/brittle_star_project/evaluation/policy.py [hunk 11]
```diff
@@ -101,11 +102,25 @@ def from_checkpoint(
...
+
     def act(self, *, observations: dict[str, Any]) -> np.ndarray:
         """Return deterministic action (actor mean, no exploration noise)."""
         batched_obs = jax.tree.map(lambda x: jnp.asarray(x)[None, ...], observations)
-        obs = self._obs_processor(batched_obs)[0]
-        hidden = self._sensor_apply(self._params["sensor_params"], obs)
-        mean, _log_std = self._actor_apply(self._params["actor_params"], hidden)
+        obs = self._obs_processor(batched_obs)
+
+        # TODO: message passing
```
Comment (tdpeuter) [inline src/brittle_star_project/evaluation/policy.py]: What is the status of this TODO?

File: scripts/simulate.sh
Comment (tdpeuter) [inline scripts/simulate.sh:1]: Please put this in scripts/ instead of in the repository root.

**Review by tdpeuter — APPROVED** [review]
I did not test the code locally, but it looks fine visually and the CI/CD tests pass so I approve 👍 

Minor questions and suggestions below, but I am setting this as approve to not block the PR any longer.

Thanks for the hard work 🥲

File: scripts/train.py (hunk 10 above)
Comment (RobinMeersman) [reply scripts/train.py]: I don't think there is. The logging config does not have a field i think to override the level (i could be wrong, was in the middle of an annoying bug, so did not look too closely)

File: src/brittle_star_project/evaluation/policy.py (hunk 11 above)
Comment (RobinMeersman) [reply src/brittle_star_project/evaluation/policy.py]: Nice catch, forgot to remove this one.

Comment (RobinMeersman) [reply src/brittle_star_project/evaluation/policy.py]: Wait no, this is not mine... I'll ask jona about this one (nothing was communicated about this)
